### ISSUE TITLE
[Question] How to slice observation in custom network?

### ISSUE THREAD

**Issue by EddieMataEwy** [DLR-RM/stable-baselines3#1487]
### ❓ Question

In the network I need to be able to slice the observation with [initial_index:final_index] (or similar like torch.tensor.select) to then process it separately from the rest of the observation.
What gets passed to the network is the observationS.
I don't understand this very well, but my guess is that this is because it is a vectorized environment.

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is minimal and working
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Hello,
could you provide more context on what and why you're trying to do?

**Comment by EddieMataEwy** [comment]
My environment has an observation space of let's say:
`self.observation_space = spaces.Box(low=-1000, high=1000, shape=(N_OBSERVATION_SPACE,), dtype=np.float32)`
This observation is then fed to the custom features extractor:
`class CustomCNN(BaseFeaturesExtractor):
    ...
    def forward(self, observations: th.Tensor) -> th.Tensor:
        partial_view = (observations[0][:600]).reshape((15, 4, 10))
        partial_cnn = self.partial_process(partial_view)
        action_view = (observations[0][600:800]).reshape((8, 5, 5))
        action_cnn = self.action_process(action_view)
        return th.concat((partial_cnn , action_cnn , observations[0][800:]))
`

This seems to work but I am concerned with having to do `observations[0]` before slicing. I feel like it should be a direct slice in the form of `observations[:600]`.

I am still testing my code but this seemed suspicious enough to want to ask about it.

**Comment by araffin** [comment]
> This seems to work but I am concerned with having to do observations[0] before slicing. 

It sounds like you are misunderstanding what is the batch dimension, this is more a PyTorch question.
In short, for efficiency, the input to the neural network contains a batch dimension, so that the inputs are not processed one by one, but several at a time.
When collecting data with a single env, the dimension will be one but when doing the gradient update, the size will be of `batch_size`.

In your case, you should probably use dictionary observation, see doc: https://stable-baselines3.readthedocs.io/en/master/guide/custom_policy.html#multiple-inputs-and-dictionary-observations

**Comment by EddieMataEwy** [comment]
Thanks a lot. I completely missed that.
That should solve it, thank you!
